diff -rpuN audio_example_generated/audio_iplayer.h audio_example_edited/audio_iplayer.h
--- audio_example_generated/audio_iplayer.h	2011-03-21 19:56:31.055896997 +0100
+++ audio_example_edited/audio_iplayer.h	2011-03-21 19:57:36.317896996 +0100
@@ -12,7 +12,7 @@ G_BEGIN_DECLS
 
 /* UserCode header_top { */
 
-    /* add further includes.../
+#include "audio_format.h"
 
 /* } UserCode */
 
diff -rpuN audio_example_generated/audio_ogg_player.c audio_example_edited/audio_ogg_player.c
--- audio_example_generated/audio_ogg_player.c	2011-03-21 19:56:31.075896997 +0100
+++ audio_example_edited/audio_ogg_player.c	2011-03-21 19:57:36.339896996 +0100
@@ -7,7 +7,7 @@
 
 /* UserCode source_top { */
 
-/* add further includes ...*/
+#include "audio_error.h"
 
 /* } UserCode */
 
@@ -140,8 +140,9 @@ audio_oggplay_instance_init(
     /* UserCode instance_init { */
     
     /* init members, allocate memory, etc ... */
-    
-    
+    self->priv->current = NULL;
+    self->priv->status = AUDIO_PLAYER_STATUS_READY;
+
     /* } UserCode */
     
 }
@@ -231,7 +232,10 @@ audio_oggplay_finalize(GObject* obj){
     
     /* UserCode finalize { */
     
-        /* free allocated memory... */
+    AudioOggPlayer* self = AUDIO_OGG_PLAYER(obj);
+
+    if (self->priv->current)
+    	g_free(self->priv->current);
     
     /* } UserCode */
     
@@ -273,8 +277,7 @@ audio_oggplay_get_property(
         
         case PROP_STATUS:
             /* UserCode property_get_status { */
-            
-            
+            g_value_set_enum(value, self->priv->status);
             /* } UserCode */
             break;
         
@@ -296,6 +299,44 @@ audio_oggplay_start_im(AudioIPlayer* obj
     GError** error) {
 /* UserCode IPlayer_start { */
 
+	AudioOggPlayer* self = AUDIO_OGG_PLAYER(obj);
+	gboolean resumed = FALSE;
+
+	if (format != AUDIO_FORMAT_OGG_VORBIS) {
+
+		if (error) {
+			*error = g_error_new(
+					AUDIO_ERROR,
+					AUDIO_ERROR_UNKNOWN_FORMAT,
+					"audio format of file \"%s\" unknown",
+					track_uri
+					);
+		}
+
+		return;
+
+	}
+
+	switch (self->priv->status) {
+	case AUDIO_PLAYER_STATUS_READY:
+		if (self->priv->current)
+			g_free(self->priv->current);
+		self->priv->current = g_strdup(track_uri);
+		self->priv->status = AUDIO_PLAYER_STATUS_PLAYING;
+		break;
+	default:
+		if (error) {
+			*error = g_error_new(
+					AUDIO_ERROR,
+					AUDIO_ERROR_INVALID_STATUS,
+					"player is not ready"
+					);
+		}
+		return;
+	}
+
+	g_signal_emit_by_name(self, "started", self->priv->current, resumed);
+
 /* } UserCode */
 }
 
@@ -303,6 +344,24 @@ static gboolean
 audio_oggplay_pause_im(AudioIPlayer* obj) {
 /* UserCode IPlayer_pause { */
 
+	AudioOggPlayer* self = AUDIO_OGG_PLAYER(obj);
+	gboolean resumed;
+
+	switch (self->priv->status) {
+	case AUDIO_PLAYER_STATUS_READY:
+		/* nothing to do */
+		break;
+	case AUDIO_PLAYER_STATUS_PLAYING:
+		self->priv->status = AUDIO_PLAYER_STATUS_PAUSED;
+		g_signal_emit_by_name(self, "paused", self->priv->current);
+		break;
+	case AUDIO_PLAYER_STATUS_PAUSED:
+		self->priv->status = AUDIO_PLAYER_STATUS_PLAYING;
+		resumed = TRUE;
+		g_signal_emit_by_name(self, "started", self->priv->current, resumed);
+		break;
+	}
+
 /* } UserCode */
 }
 
@@ -310,6 +369,21 @@ static gboolean
 audio_oggplay_stop_im(AudioIPlayer* obj) {
 /* UserCode IPlayer_stop { */
 
+	AudioOggPlayer* self = AUDIO_OGG_PLAYER(obj);
+	gchar* tmp = NULL;
+
+	if (self->priv->current) {
+		tmp = self->priv->current;
+		self->priv->current = NULL;
+	}
+
+	self->priv->status = AUDIO_PLAYER_STATUS_READY;
+
+	g_signal_emit_by_name(self, "stopped", tmp);
+
+	if (tmp)
+		g_free(tmp);
+
 /* } UserCode */
 }
 
diff -rpuN audio_example_generated/audio_ogg_player.h audio_example_edited/audio_ogg_player.h
--- audio_example_generated/audio_ogg_player.h	2011-03-21 19:56:31.072896997 +0100
+++ audio_example_edited/audio_ogg_player.h	2011-03-21 19:57:36.335896996 +0100
@@ -14,7 +14,7 @@ G_BEGIN_DECLS
 
 /* UserCode header_top { */
 
-    /* add includes here... */
+#include "audio_player_status.h"
 
 /* } UserCode */
 
diff -rpuN audio_example_generated/demo.c audio_example_edited/demo.c
--- audio_example_generated/demo.c	1970-01-01 01:00:00.000000000 +0100
+++ audio_example_edited/demo.c	2011-03-20 22:47:08.301231858 +0100
@@ -0,0 +1,95 @@
+#include "audio_ogg_player.h"
+
+void
+on_started(AudioIPlayer* object, const gchar* track_uri, gboolean resumed,
+		gpointer user_data);
+
+void
+on_stopped(AudioIPlayer* object, const gchar* track_uri, gpointer user_data);
+
+void
+on_paused(AudioIPlayer* object, const gchar* track_uri, gpointer user_data);
+
+void log_error(GError** error);
+
+int main() {
+
+	AudioIPlayer* player;
+	gchar* stream_info = "Deutschlandradio Kultur";
+	GError* error = NULL;
+
+	g_type_init();
+
+	player = AUDIO_IPLAYER(audio_oggplay_new());
+
+	g_signal_connect(player, "started", G_CALLBACK(on_started), stream_info);
+	g_signal_connect(player, "stopped", G_CALLBACK(on_stopped), stream_info);
+	g_signal_connect(player, "paused", G_CALLBACK(on_paused), stream_info);
+
+	audio_iplayer_start(player, "bird_of_paradise.ogg",
+			AUDIO_FORMAT_OGG_VORBIS, &error);
+
+	if (error)
+		log_error(&error);
+
+	audio_iplayer_stop(player);
+
+	audio_iplayer_start(player, "relaxin_at_camarillo.ogg",
+			AUDIO_FORMAT_OGG_VORBIS, &error);
+
+	if (error)
+		log_error(&error);
+
+	audio_iplayer_pause(player);
+
+	audio_iplayer_pause(player);
+
+	audio_iplayer_stop(player);
+
+	audio_iplayer_start(player, "moondreams.mp3", AUDIO_FORMAT_MP3, &error);
+
+	if (error)
+		log_error(&error);
+
+	g_object_unref(player);
+
+	return 0;
+
+}
+
+void on_started(AudioIPlayer* object, const gchar* track_uri, gboolean resumed,
+		gpointer user_data) {
+
+	gchar* stream_info = (gchar*) user_data;
+
+	if (!resumed)
+		g_print("(%s) Now playing: %s\n", stream_info, track_uri);
+	else
+		g_print("(%s) Resuming: %s\n", stream_info, track_uri);
+
+}
+
+void on_stopped(AudioIPlayer* object, const gchar* track_uri,
+		gpointer user_data) {
+
+	gchar* stream_info = (gchar*) user_data;
+
+	g_print("(%s) Stopped: %s\n", stream_info, track_uri);
+
+}
+
+void on_paused(AudioIPlayer* object, const gchar* track_uri, gpointer user_data) {
+
+	gchar* stream_info = (gchar*) user_data;
+
+	g_print("(%s) Paused: %s\n", stream_info, track_uri);
+
+}
+
+void log_error(GError** error) {
+
+	g_printerr("%s\n", (*error)->message);
+	g_error_free(*error);
+	*error = NULL;
+
+}
diff -rpuN audio_example_generated/Makefile audio_example_edited/Makefile
--- audio_example_generated/Makefile	1970-01-01 01:00:00.000000000 +0100
+++ audio_example_edited/Makefile	2011-03-20 19:36:03.044190479 +0100
@@ -0,0 +1,35 @@
+CFLAGS=`pkg-config gobject-2.0 --cflags`
+LIBS=`pkg-config gobject-2.0 --libs`
+
+OBJ_FILES=\
+audio_format.o \
+audio_iplayer.o \
+audio_iplayer_marshaller.o \
+audio_ogg_player.o \
+audio_player_status.o \
+demo.o
+
+audio_format.o: audio_format.c
+	gcc -c $(CFLAGS) audio_format.c	
+
+audio_iplayer.o: audio_iplayer.c
+	gcc -c $(CFLAGS) audio_iplayer.c
+
+audio_iplayer_marshaller.o: audio_iplayer_marshaller.c
+	gcc -c $(CFLAGS) audio_iplayer_marshaller.c	
+	
+audio_ogg_player.o: audio_ogg_player.c
+	gcc -c $(CFLAGS) audio_ogg_player.c	
+
+audio_player_status.o: audio_player_status.c
+	gcc -c $(CFLAGS) audio_player_status.c	
+
+demo.o: demo.c
+	gcc -c $(CFLAGS) demo.c	
+
+demo: $(OBJ_FILES)
+	gcc -o demo $(LIBS) $(OBJ_FILES) 
+
+clean:
+	rm -f *.o
+	rm -f demo
